### PR TITLE
<utility> P0586R2 Integer Comparison Functions

### PR DESCRIPTION
**Pull request by Neargye**
Description
===========

Will resolve issue #560

Checklist
=========

- [x] Implement P0586R2 
- [x] Add __cpp_lib_integer_comparison_functions 
- [x] Recheck P0586R2 

Be sure you've read README.md and understand the scope of this repo.

If you're unsure about a box, leave it unchecked. A maintainer will help you.

- [x] Identifiers in product code changes are properly `_Ugly` as per
  https://eel.is/c++draft/lex.name#3.1 or there are no product code changes.
- [x] The STL builds successfully and all tests have passed (must be manually
  verified by an STL maintainer before automated testing is enabled on GitHub,
  leave this unchecked for initial submission).
- [x] These changes introduce no known ABI breaks (adding members, renaming
  members, adding virtual functions, changing whether a type is an aggregate
  or trivially copyable, etc.).
- [x] These changes were written from scratch using only this repository,
  the C++ Working Draft (including any cited standards), other WG21 papers
  (excluding reference implementations outside of proposed standard wording),
  and LWG issues as reference material. If they were derived from a project
  that's already listed in NOTICE.txt, that's fine, but please mention it.
  If they were derived from any other project (including Boost and libc++,
  which are not yet listed in NOTICE.txt), you *must* mention it here,
  so we can determine whether the license is compatible and what else needs
  to be done.

### CODE REVIEW

File: stl/inc/utility [hunk 1]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+
+template <typename _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
```
Comment (miscco) [inline stl/inc/utility]: I think it would be better to make this a concept
```suggestion
concept _Integral_not_bool_char = is_integral_v<_Ty>
```

File: stl/inc/utility [hunk 2]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+#if _HAS_CXX20
+
+template <typename _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
+    && !is_same_v<_Ty, bool> && !is_same_v<_Ty, char>
+    && !is_same_v<_Ty, char8_t> && !is_same_v<_Ty, char16_t>
+    && !is_same_v<_Ty, char32_t> && !is_same_v<_Ty, wchar_t>;
+
+// SAFE INTEGRAL COMPARISONS
+template <typename _Ty1, typename _Ty2>
```
Comment (miscco) [inline stl/inc/utility]: Then this could be 
```suggestion
template <_Integral_not_bool_char _Ty1, _Integral_not_bool_char _Ty2>

```

File: stl/inc/utility [hunk 3]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
```
Comment (miscco) [inline stl/inc/utility]: That would be superfluous. Unfortunately the paper is not yet there so I can only guess but my guess would be that this is a `Mandates:` not an `Expects:`. The former requires that the function is not part of the overload set.

Do you have the paper to confirm what the standard requires?

File: stl/inc/utility [hunk 4]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Value <= (numeric_limits<_Ty1>::max)();
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Value >= 0 && _Value <= (numeric_limits<_Ty1>::max)();
+    } else {
+        return (numeric_limits<_Ty1>::min)() <= _Value && _Value <= (numeric_limits<_Ty1>::max)();
+    }
+}
+
+template <typename _Ty1, typename _Ty2>
```
Comment (miscco) [inline stl/inc/utility]: same here

File: stl/inc/utility [hunk 5]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Left >= 0 || static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right;
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Right >= 0 && _Left == static_cast<make_unsigned_t<_Ty2>>(_Right);
+    } else {
+        return _Left == _Right;
+    }
+}
+
+template <typename _Ty1, typename _Ty2>
```
Comment (miscco) [inline stl/inc/utility]: And everywhere else

File: stl/inc/utility [hunk 6]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+constexpr bool cmp_not_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    return !cmp_equal(_Left, _Right);
```
Comment (miscco) [inline stl/inc/utility]: I think @BillyONeal cares about debug performance so I would guess you have to open-code it.

That said these comparison functions are so hairy that you might be lucky (Please wait for a maintainer to confirm / reject before you do any work)

File: stl/inc/utility [hunk 7]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+    && !is_same_v<_Ty, bool> && !is_same_v<_Ty, char>
+    && !is_same_v<_Ty, char8_t> && !is_same_v<_Ty, char16_t>
+    && !is_same_v<_Ty, char32_t> && !is_same_v<_Ty, wchar_t>;
+
+// SAFE INTEGRAL COMPARISONS
+template <typename _Ty1, typename _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Value <= (numeric_limits<_Ty1>::max)();
```
Comment (Svido) [inline stl/inc/utility]: This is a signed / unsigned mismatch.
You can cast numeric_limit<_Ty1>::max() to make_unsiged<_Ty1> to silence compiler warnings.

Note that it's possible to base the implementation on the other function templates.
`return cmp_greater_equal(_Value, (numeric_limits<_Ty1>::min()) && (cmp_less_equal(_Value, numeric_limits<_Ty1>::max());`

File: stl/inc/utility [hunk 8]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Value >= 0 && _Value <= (numeric_limits<_Ty1>::max)();
```
Comment (Svido) [inline stl/inc/utility]: Also a signed / unsigned mismatch.

File: stl/inc/utility [hunk 9]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
...
+    } else {
+        return (numeric_limits<_Ty1>::min)() <= _Value && _Value <= (numeric_limits<_Ty1>::max)();
+    }
+}
+
+template <typename _Ty1, typename _Ty2>
+constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Left >= 0 || static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right;
```
Comment (Svido) [inline stl/inc/utility]: This seems to be incorrect. If _Left is >= 0 this is always true. So cmp_equal(1, 42) is true.
Replacing || with && is probably what you meant.

File: stl/inc/utility (hunk 3 above)
Comment (Svido) [reply stl/inc/utility]: The paper is not public available, but it's already merged in the current working draft repo. It is a mandates clause, as you can find [here](https://github.com/cplusplus/draft/blob/d34fb9a542c3a9873e49e53a8a86eccfdd336109/source/utilities.tex#L546-L548).

File: stl/inc/utility (hunk 9 above)
Comment (Neargye) [reply stl/inc/utility]: Yes, it looks like an error. I'm still working on tests, so I’ll fix all the bugs soon.

File: stl/inc/utility (hunk 1 above)
Comment (Neargye) [reply stl/inc/utility]: Yes, I forgot about the concepts. I still used C++17.

File: stl/inc/utility (hunk 3 above)
Comment (Neargye) [reply stl/inc/utility]: I based on [draft](https://eel.is/c++draft/utility.intcmp) and [p0586r1](http://open-std.org/JTC1/SC22/WG21/docs/papers/2018/p0586r1.html)

```
For each of the following functions, if either of `T` or `U` is not
  a standard integer type or an extended integer type, as specified in 6.9.1,
  the call is ill-formed.
  [Note: std::byte, char, char16_t, char32_t, wchar_t, and bool are not
         comparable with these functions. --end note]
```

File: stl/inc/utility [hunk 10]
```diff
@@ -644,6 +645,77 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+
+template <typename _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
+    && !is_same_v<_Ty, bool> && !is_same_v<_Ty, char>
+    && !is_same_v<_Ty, char8_t> && !is_same_v<_Ty, char16_t>
+    && !is_same_v<_Ty, char32_t> && !is_same_v<_Ty, wchar_t>;
```
Comment (CaseyCarter) [inline stl/inc/utility]: We can shorten this to:
```suggestion
template <class _Ty>
inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
    && !_Is_any_of_v<_Ty, bool, char, 
#ifdef __cpp_char8_t
        char8_t, 
#endif // __cpp_char8_t
        char16_t, char32_t, wchar_t>;
```

Comment (miscco) [reply stl/inc/utility]: Counts the number of lines o.O

But even if its longer It seems more correct given that `char8_t `might not be defined

Comment (BillyONeal) [reply stl/inc/utility]: I think something like the following would be best for the hyper-throughput-paranoid, that's how we did _Is_character in <xutility>:

```
template<class _Ty>
inline constexpr bool _Is_integral_non_bool_char = is_integral_v<_Ty>;
template<>
inline constexpr bool _Is_integral_non_bool_char<char> = false;
template<>
inline constexpr bool _Is_integral_non_bool_char<char16_t> = false;
// repeat for each
```

Comment (Neargye) [reply stl/inc/utility]: Should concepts be used here? @CaseyCarter

File: stl/inc/utility (hunk 3 above)
Comment (Neargye) [reply stl/inc/utility]: P0586R2 now available http://open-std.org/JTC1/SC22/WG21/docs/papers/2020/p0586r2.html

File: stl/inc/utility (hunk 8 above)
Comment (Neargye) [reply stl/inc/utility]: fix in new commit

File: stl/inc/utility (hunk 7 above)
Comment (Neargye) [reply stl/inc/utility]: fix in new commit

File: stl/inc/utility (hunk 9 above)
Comment (Neargye) [reply stl/inc/utility]: Thanks! fix in new commit.

File: stl/inc/utility (hunk 1 above)
Comment (miscco) [reply stl/inc/utility]: I dont know about the plans for moving the requirement up to clang-10 which would benecessary to support concepts.

That said if you do go with concepts you should use them in the functions so that the function is SFINAED out.
Otherwise it should be `enable_if_t`

File: stl/inc/utility [hunk 11]
```diff
@@ -644,6 +645,82 @@ template <size_t _Idx>
...
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// SAFE INTEGRAL COMPARISONS
+template <class _Ty1, class _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
```
Comment (CaseyCarter) [inline stl/inc/utility]: I think I'd rather see `!is_signed_v` than `is_unsigned_v` in all of these functions - there's no need to instantiate four variable templates when two would do. (Equivalently, you could reorder your cases as the working draft does:
```c++
if constexpr (both are signed) {}
else if constexpr (left is signed) {}
else {}
```
which I would mildly prefer, but that's simply style preference so your call.)

File: stl/inc/utility [hunk 12]
```diff
@@ -644,6 +645,82 @@ template <size_t _Idx>
...
+    } else {
+        return (numeric_limits<_Ty1>::min)() <= _Value && _Value <= (numeric_limits<_Ty1>::max)();
+    }
+}
+
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    if constexpr (is_signed_v<_Ty1> && is_unsigned_v<_Ty2>) {
+        return _Left >= 0 && static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right;
```
Comment (CaseyCarter) [inline stl/inc/utility]: I think it would make sense to reorder the clauses in all of these `meow >= 0 && convert_to_unsigned(meow) == woof` expressions as `convert_to_unsigned(meow) == woof && meow >= 0`. The latter form rejects more `false` cases - those in which `meow` is positive but differs from `woof` - earlier.

File: stl/inc/utility [hunk 13]
```diff
@@ -644,6 +645,82 @@ template <size_t _Idx>
...
+    } else if constexpr (is_unsigned_v<_Ty1> && is_signed_v<_Ty2>) {
+        return _Right >= 0 && _Left == static_cast<make_unsigned_t<_Ty2>>(_Right);
+    } else {
+        return _Left == _Right;
+    }
+}
+
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_not_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
```
Comment (CaseyCarter) [inline stl/inc/utility]: This is redundant with the check in `cmp_equal` (And on 706, 712, and 718).
```suggestion
```

File: stl/inc/utility [hunk 14]
```diff
@@ -644,6 +645,82 @@ template <size_t _Idx>
...
+    return !cmp_less(_Right, _Left);
+}
+
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
+    return !cmp_less(_Left, _Right);
+}
+
```
Comment (CaseyCarter) [inline stl/inc/utility]: Ditto "no blank line padding inside conditional"
```suggestion
```

File: stl/inc/utility [hunk 15]
```diff
@@ -644,6 +645,82 @@ template <size_t _Idx>
...
+#endif // _NATIVE_WCHAR_T_DEFINED
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// SAFE INTEGRAL COMPARISONS
+template <class _Ty1, class _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>);
```
[the remark this answers is not in the document]
Comment (miscco) [reply stl/inc/utility]: But is this conforming given that it is a constraint?

Shouldnt that be SFINAEd out?

[the remark this answers is not in the document]
Comment (CaseyCarter) [reply stl/inc/utility]: This isn't implementing a _Constraints_ element, it's implementing a _Mandates_ element (http://eel.is/c++draft/utility.intcmp#9).

File: stl/inc/utility [hunk 16]
```diff
@@ -644,6 +645,79 @@ template <size_t _Idx>
...
+    && !_Is_any_of_v<_Ty, bool, char,
+#ifdef _NATIVE_WCHAR_T_DEFINED
+        wchar_t,
+#endif // _NATIVE_WCHAR_T_DEFINED
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// SAFE INTEGRAL COMPARISONS
```
Comment (CaseyCarter) [inline stl/inc/utility:660]: This is not a conventional SHOUTY BANNER. Per our shouty banner convention this should be `// FUNCTION TEMPLATE cmp_equal` (and similarly for the other function templates). (Let the record show that I hate the shouty banners, but it's still the convention until we can resolve #306 in the name of all that's right and good.)

**Comment by Neargye** [comment]
@CaseyCarter update pr.
Will add test soon to `tests/std/tests/P0586R2_integer_comparison`

File: stl/inc/yvals_core.h [hunk 17]
```diff
@@ -1105,26 +1106,27 @@
...
-#define __cpp_lib_to_array                 201907L
-#define __cpp_lib_type_identity            201806L
-#define __cpp_lib_unwrap_ref               201811L
+#define __cpp_lib_constexpr_algorithms         201806L
+#define __cpp_lib_constexpr_memory             201811L
+#define __cpp_lib_constexpr_numeric            201911L
+#define __cpp_lib_endian                       201907L
+#define __cpp_lib_erase_if                     202002L
+#define __cpp_lib_generic_unordered_lookup     201811L
+#define __cpp_lib_integer_comparison_functions 202002L
```
Comment (StephanTLavavej) [inline stl/inc/yvals_core.h:1115]: `tests/std/tests/VSO_0157762_feature_test_macros/test.cpp` needs to be updated accordingly.

File: stl/inc/utility [hunk 18]
```diff
@@ -644,6 +644,88 @@ template <size_t _Idx>
...
+        wchar_t,
+#endif // _NATIVE_WCHAR_T_DEFINED
+#ifdef __cpp_char8_t
+        char8_t,
+#endif // __cpp_char8_t
+        char16_t, char32_t>;
+
+// FUNCTION TEMPLATE cmp_equal
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
```
Comment (StephanTLavavej) [inline stl/inc/utility:661]: As pure observers, I believe that all of these functions should be marked as `_NODISCARD`.

File: stl/inc/utility [hunk 19]
```diff
@@ -644,6 +644,88 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+template <class _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
```
Comment (StephanTLavavej) [inline stl/inc/utility:649]: Naming: We currently have `_Is_nonbool_integral`. I think that `not_bool_char`, in addition to being inconsistent with `nonbool`, is also somewhat confusing - this type trait rejects both `char` and `char16_t` etc. but permits `signed char` and `unsigned char`.

I recommend imitating the Standardese and calling this `_Is_standard_integer`. (We can ignore the "extended" bit since we don't have any such types.) Then this would be consistent with:
https://github.com/microsoft/STL/blob/43e07f3fad934b710d2de21ffdcf4cd72267c826/stl/inc/bit#L39-L41

File: stl/inc/utility [hunk 20]
```diff
@@ -644,6 +644,88 @@ template <size_t _Idx>
 inline constexpr in_place_index_t<_Idx> in_place_index{};
 #endif // _HAS_CXX17
 
+#if _HAS_CXX20
+template <class _Ty>
+inline constexpr bool _Is_integral_not_bool_char = is_integral_v<_Ty>
+    && !_Is_any_of_v<_Ty, bool, char,
```
Comment (StephanTLavavej) [inline stl/inc/utility:650]: Semantics: `is_integral_v`, being a primary type category, ignores top-level cv-qualifiers. This `_Is_any_of_v` check, being an extended `is_same` trait, is sensitive to top-level cv-qualifiers. Although it may not matter given your specific usage below, in the past we've found that internal type traits not following Standard type trait conventions can lead to headaches. I recommend using `remove_cv_t<_Ty>` here. (The throughput cost is very low, as the compiler optimizes that trait.)

File: stl/inc/utility (hunk 6 above)
Comment (StephanTLavavej) [reply stl/inc/utility:676]: I tend to think that this is fine as-is; it's true that it pays one more function call for debug perf/codegen size, but I don't expect these to be truly hammered by applications, and any application that really does hammer this could be writing the comparisons directly (carefully). As you observed, there is a significant cost to replicating these complicated functions with significant mutations. We can always change this code later if we receive reported issues.

(My opinion matters the least here, as I am historically the least sensitive to debug perf concerns, although I am trying to reform :cat: )

File: stl/inc/utility [hunk 21]
```diff
@@ -644,6 +644,88 @@ template <size_t _Idx>
...
+        return _Left == static_cast<make_unsigned_t<_Ty2>>(_Right) && _Right >= 0;
+    } else {
+        return static_cast<make_unsigned_t<_Ty1>>(_Left) == _Right && _Left >= 0;
+    }
+}
+
+// FUNCTION TEMPLATE cmp_not_equal
+template <class _Ty1, class _Ty2>
+constexpr bool cmp_not_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !cmp_equal(_Left, _Right);
```
Comment (StephanTLavavej) [inline stl/inc/utility:676]: This non-`_Ugly` function call should be `_STD` qualified, and this issue occurs below. (You ultimately have a `static_assert` restricting this to standard integer types, but (1) ADL hijacking might happen before that is reached, and (2) our convention is to `_STD` qualify all non-`_Ugly` function calls even when they're theoretically safe.)

File: stl/inc/utility [hunk 22]
```diff
@@ -644,6 +644,88 @@ template <size_t _Idx>
...
+}
+
+// FUNCTION TEMPLATE in_range
+template <class _Ty1, class _Ty2>
+constexpr bool in_range(const _Ty2 _Value) noexcept {
+    static_assert(_Is_integral_not_bool_char<_Ty1> && _Is_integral_not_bool_char<_Ty2>,
+        "The integer comparison functions only accept standard and extended integer types.");
+    using _UTy1 = make_unsigned_t<_Ty1>;
+    // _U_max is numeric_limits<_Ty1>::max() converted to _UTy1
+    constexpr auto _U_max = static_cast<_UTy1>(static_cast<_UTy1>(-1) >> is_signed_v<_Ty1>);
```
Comment (StephanTLavavej) [inline stl/inc/utility:718]: Right-shifting by a `bool` promoted to `int` is correct, but unusual. Should we `static_cast<int>` here? (I'm torn between "don't help the compiler" and "make weird code look distinctive". I suppose this is fine as-is.)

Comment (StephanTLavavej) [inline stl/inc/utility:718]: I wonder if this should contain `constexpr` programming to return guaranteed answers when we can statically see that the bounds are safe. When `_Ty1`'s range is a (non-strict) superset of `_Ty2`'s range (including when they're the same type), we can always return `true`. We can also skip bounds checks in some cases - e.g. when `is_signed_v<_Ty2>` but `_Ty1` is unsigned, the code below performs two checks, but at constexpr time we can detect if `_Ty1`'s upper bound is plenty - e.g. `_Ty2` could be `int16_t` and `_Ty1` could be `uint64_t` so we only need the `>= 0` check.

So, I think I'm asking for a significant refactoring. I haven't worked out what the code should look like, but I believe it would consider each bound (lower/upper) separately, and use constexpr tests to determine if they're always satisfied, or if we need a runtime comparison, leading to `return true;` in cases where the bounds are definitely ok.

This would also avoid compiler warnings when comparisons are always true.

Comment (StephanTLavavej) [reply stl/inc/utility:718]: Also, if such logic would be really obnoxious, then it's fine to not do this - I'm just wondering whether it's possible to avoid compiler warnings and improve debug codegen at minimal cost.

File: stl/inc/utility (hunk 21 above)
Comment (Neargye) [reply stl/inc/utility:676]: Fix in a new commit, could you please check?

File: stl/inc/utility (hunk 22 above)
Comment (StephanTLavavej) [reply stl/inc/utility:721]: This triggered `warning C4804: '>>': unsafe use of type 'bool' in operation`. I'll push a fix.

**Comment by StephanTLavavej** [comment]
Reviewing and editing. I pushed an initial batch of changes:

* Fix typo in `VSO_0157762_feature_test_macros`.
* Need to mention new test in `tests/std/test.lst`.
* test.cpp cleanups.
  + We aren't including `<Windows.h>` so we don't need `NOMINMAX`.
  + This test uses `usual_latest_matrix.lst` so we can assume that the feature is always available.
  + Consistently use `const`.
* Unify compile-time and run-time testing.
* Fix `warning C4804: '>>': unsafe use of type 'bool' in operation`
* Fix test failure. In MSVC's LLP64 ABI, `int` and `long` are both 32-bit.

File: stl/inc/utility (hunk 21 above)
Comment (StephanTLavavej) [reply stl/inc/utility]: Yep, looks good!

File: stl/inc/utility [hunk 23]
```diff
@@ -647,6 +647,88 @@ template <size_t _Idx>
...
+}
+
+// FUNCTION TEMPLATE cmp_greater_equal
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !_STD cmp_less(_Left, _Right);
+}
+
+// FUNCTION TEMPLATE in_range
+template <class _Ty1, class _Ty2>
```
Comment (CaseyCarter) [inline stl/inc/utility:715]: The Standard names these parameters `R` and `T`; should we similarly call them `_Result`  and `_Ty`?

File: tests/std/tests/P0586R2_integer_comparison/test.cpp [hunk 24]
```diff
@@ -0,0 +1,106 @@
...
+    assert(!in_range<int>(numeric_limits<long long>::max()));
+    assert(!in_range<int>(numeric_limits<long long>::min()));
+    assert(in_range<int>(numeric_limits<int>::max()));
+    assert(in_range<int>(numeric_limits<int>::min()));
+    assert(in_range<long>(numeric_limits<int>::max()));
+    assert(in_range<long>(numeric_limits<int>::min()));
+
+    { // Different signs
+        const int i          = -1;
+        const unsigned int u = 1;
```
Comment (CaseyCarter) [inline tests/std/tests/P0586R2_integer_comparison/test.cpp:59]: I'd like to see some `make_signed_t<T>{-1}, make_unsigned_t<T>(-1)` test cases as well.

File: stl/inc/utility (hunk 23 above)
Comment (StephanTLavavej) [reply stl/inc/utility:715]: Great catch. It's unclear to me whether `R` means `Result` or `Range`. (The return type of this function is `bool`, after all.) I propose renaming them to `_Rx` and `_Ty`, which will align more closely with the Standard, look more visually distinct, and will avoid selecting a particular interpretation for `R`.

File: stl/inc/utility (hunk 22 above)
Comment (StephanTLavavej) [reply stl/inc/utility]: I figured out a really elegant way of expressing this. Will push a change.

File: stl/inc/utility [hunk 25]
```diff
@@ -712,20 +712,30 @@ _NODISCARD constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right)
...
-    // _U_max is numeric_limits<_Ty1>::max() converted to _UTy1
-    constexpr auto _U_max = static_cast<_UTy1>(static_cast<_UTy1>(-1) >> static_cast<int>(is_signed_v<_Ty1>));
-    if constexpr (is_signed_v<_Ty1> == is_signed_v<_Ty2>) {
-        return static_cast<_Ty1>(_U_max + 1) <= _Value && _Value <= static_cast<_Ty1>(_U_max);
-    } else if constexpr (is_signed_v<_Ty2>) {
-        return _Value >= 0 && static_cast<make_unsigned_t<_Ty2>>(_Value) <= static_cast<_Ty1>(_U_max);
-    } else {
-        return _Value <= _U_max;
+
+    constexpr auto _Ty_min = (numeric_limits<_Ty>::min)();
```
[the remark this answers is not in the document]
Comment (StephanTLavavej) [reply stl/inc/utility:720]: Oops! Sorry for damaging the code. That’s what I deserve for saving time by running just the one test, instead of a full test pass. I’ll fix it 😸

File: stl/inc/utility [hunk 26]
```diff
@@ -647,6 +647,120 @@ template <size_t _Idx>
...
+
+// FUNCTION TEMPLATE cmp_greater_equal
+template <class _Ty1, class _Ty2>
+_NODISCARD constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right) noexcept {
+    return !_STD cmp_less(_Left, _Right);
+}
+
+// FUNCTION TEMPLATE in_range
+template <class _Ty>
+_NODISCARD constexpr _Ty _Min_limit() noexcept { // same as (numeric_limits<_Ty>::min)(), less throughput cost
```
Comment (AdamBucior) [inline stl/inc/utility:716]: Why not make it `_CONSTEVAL`?

File: stl/inc/utility [hunk 27]
```diff
@@ -647,6 +647,120 @@ template <size_t _Idx>
...
+    if constexpr (is_signed_v<_Ty>) {
+        constexpr auto _Unsigned_max = static_cast<make_unsigned_t<_Ty>>(-1);
+        return static_cast<_Ty>((_Unsigned_max >> 1) + 1); // well-defined, N4861 [conv.integral]/3
+    } else {
+        return 0;
+    }
+}
+
+template <class _Ty>
+_NODISCARD constexpr _Ty _Max_limit() noexcept { // same as (numeric_limits<_Ty>::max)(), less throughput cost
```
Comment (AdamBucior) [inline stl/inc/utility:727]: Ditto

File: stl/inc/utility [hunk 28]
```diff
@@ -712,21 +712,53 @@ _NODISCARD constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right)
...
+template <class _Rx, class _Ty>
+_NODISCARD constexpr bool in_range(const _Ty _Value) noexcept {
+    static_assert(_Is_standard_integer<_Rx> && _Is_standard_integer<_Ty>,
+        "The integer comparison functions only accept standard and extended integer types.");
+
+    constexpr auto _Ty_min = _Min_limit<_Ty>();
+    constexpr auto _Rx_min = _Min_limit<_Rx>();
+
+    if constexpr (_STD cmp_less(_Ty_min, _Rx_min)) {
+        if (_Value < static_cast<_Ty>(_Rx_min)) {
```
Comment (CaseyCarter) [inline stl/inc/utility:746]: `_Rx_min` is a constant expression, and its conversion to `_Ty` is non-narrowing. We typically document and enforce that with `{}` conversion: (This is probably not worth resetting testing.)
```suggestion
        if (_Value < _Ty{_Rx_min}) {
```

File: stl/inc/utility [hunk 29]
```diff
@@ -712,21 +712,53 @@ _NODISCARD constexpr bool cmp_greater_equal(const _Ty1 _Left, const _Ty2 _Right)
...
+            return false;
+        }
+    }
+
+    constexpr auto _Ty_max = _Max_limit<_Ty>();
+    constexpr auto _Rx_max = _Max_limit<_Rx>();
+
+    if constexpr (_STD cmp_greater(_Ty_max, _Rx_max)) {
+        if (_Value > static_cast<_Ty>(_Rx_max)) {
```
Comment (CaseyCarter) [inline stl/inc/utility:755]: Ditto-ish:
```suggestion
        if (_Value > _Ty{_Rx_max}) {
```

File: stl/inc/utility (hunk 26 above)
Comment (StephanTLavavej) [reply stl/inc/utility:716]: Thanks! I had simply forgotten about `_CONSTEVAL`'s existence. :joy_cat: Unfortunately, this triggered a compiler bug, which I've reduced and reported as Microsoft-internal VSO-1125263 "EDG rejects valid code using consteval":

```
C:\Temp>type kitty.cpp
#include <stdio.h>
#include <type_traits>
using namespace std;

#ifdef __cpp_consteval
template <typename T>
[[nodiscard]] consteval T MinLimit() noexcept {
    static_assert(is_integral_v<T>);
    if constexpr (is_signed_v<T>) {
        constexpr auto unsigned_max = static_cast<make_unsigned_t<T>>(-1);
        return static_cast<T>((unsigned_max >> 1) + 1);
    } else {
        return 0;
    }
}

template <typename T>
[[nodiscard]] consteval T MaxLimit() noexcept {
    static_assert(is_integral_v<T>);
    if constexpr (is_signed_v<T>) {
        constexpr auto unsigned_max = static_cast<make_unsigned_t<T>>(-1);
        return static_cast<T>(unsigned_max >> 1);
    } else {
        return static_cast<T>(-1);
    }
}

int main() {
    constexpr int min_int = MinLimit<int>();
    constexpr int max_int = MaxLimit<int>();
    printf("min_int: %d\n", min_int);
    printf("max_int: %d\n", max_int);
}
#else
    #error This test assumes that __cpp_consteval is defined.
#endif

C:\Temp>cl /EHsc /nologo /W4 /std:c++latest /c /BE kitty.cpp
kitty.cpp
"kitty.cpp", line 29: error: call to consteval function "MinLimit<T>() noexcept
          [with T=int]" did not produce a valid constant expression
      constexpr int min_int = MinLimit<int>();
                              ^
"kitty.cpp", line 29: note: cannot call non-constexpr function "MinLimit<T>()
          noexcept [with T=int]" (declared at line 7)
      constexpr int min_int = MinLimit<int>();
                              ^

"kitty.cpp", line 30: error: call to consteval function "MaxLimit<T>() noexcept
          [with T=int]" did not produce a valid constant expression
      constexpr int max_int = MaxLimit<int>();
                              ^
"kitty.cpp", line 30: note: cannot call non-constexpr function "MaxLimit<T>()
          noexcept [with T=int]" (declared at line 18)
      constexpr int max_int = MaxLimit<int>();
                              ^
```

Given that EDG is the only front-end defining `__cpp_consteval` right now, this effectively blocks us from using `_CONSTEVAL` here. (C1XX hasn't implemented it, and Clang 10's support is partial enough that they don't define the feature-test macro - although ironically, Clang 10 accepts this code if we don't query the macro.) I've recorded your excellent suggestion, so we can revisit this when one or more compilers actually accept `_CONSTEVAL` here.

File: stl/inc/utility (hunk 28 above)
Comment (StephanTLavavej) [reply stl/inc/utility]: This is a great idea, and worth resetting testing! (Realizing that converting the limit would be non-narrowing is what led to this refactoring.)

**Comment by StephanTLavavej** [comment]
Thanks again for implementing this C++20 feature! It'll ship in VS 2019 16.7 Preview 3. 😺
